pulsar-common/src/main/java/org/apache/pulsar/common/compression/ZStdRawCompressor.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.compression;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * Expose ZstdFrameCompressor which is a package protected class.
+ */
+public class ZStdRawCompressor {
+
+    private final static Method compressMethod;
+
+    static {
+        try {
+            Class<?> frameCompressor = ZStdRawCompressor.class.getClassLoader()
+                    .loadClass("io.airlift.compress.zstd.ZstdFrameCompressor");
+            compressMethod = frameCompressor.getDeclaredMethod("compress", Object.class, long.class, long.class,
+                    Object.class, long.class, long.class, int.class);
+        } catch (ClassNotFoundException | NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static int compress(long inputAddress, long inputLimit,
+                               long outputAddress, long outputLimit, int compressionLevel) {
+        try {
+            return (int) compressMethod.invoke(null, null, inputAddress,
```

Comment:
Using reflection will affect performances, do you have numbers to about the impact ?
isn't there any other way, by playing with package names ?

Reply:
In what direction were you thinking when you say 'playing with packages'? I don't see a way out at the moment.

Reply:
I took a look deeper

It looks like at airlift they are not happy to make the class public
https://github.com/airlift/aircompressor/pull/101

I wonder if we can cheat the system (at least while working in classpath mode) and created dynamically the io.airlift.compress.zstd.ZStdRawDecompressor class in the right package.

You will also need to add an interface in order to be able to call the method as you cannot bind directly to io.airlift.compress.zstd.ZStdRawDecompressor  at compile time.

The complexity of this solution is not trivial, so probably the best to do is to create some little JHM benchmark and understand how much we are going to pay for using reflection before spending time.

Did you evaluate to use MethodHandles as well ?


That's my humble opinion, I am not asking you to spend more and more time on this work,
let's wait for feedback from the others in the community